jupyter_ydoc/ydoc.py

```diff
@@ -83,6 +88,13 @@ def __init__(self, *args, **kwargs):
         self._ymeta = self._ydoc.get_map("meta")
         self._ycells = self._ydoc.get_array("cells")
 
+    def __del__(self):
+        self._ycells.unobserve(self.on_cells_changed)
+
+    def initialize(self):
```

Comment:
Why not part of `__init__`?

Reply:
Because if we observe in `__init__`, we call the callback with the changes from the initialization.
We first instaciate the YNotebook (`__init__`), and then we add the content (from disk or the updates from the store). If we observe `ycell` in the `__init__`, we will call the callback as we add updates from the store. If this store has an update that leaves the cell list empty, we will add a cell before we finish the initialization.

Reply:
Where is `initialize()` called?

Reply:
In `jupyter_server_ydoc` after adding the updates from the store or from disk

Reply:
Maybe this logic should live in `jupyter_server_ydoc`? We already [observe](https://github.com/jupyter-server/jupyter_server_ydoc/blob/d7e728b0c773977c6a307d7a8878b40c6be3a13d/jupyter_server_ydoc/ydoc.py#L171) document changes there, in order to auto-save after each change.
If we keep it here in `jupyter_ydoc`, I think we should change the name: `initialize()` doesn't say much about what it's really doing. Maybe something like `ensure_not_empty()` would be better? For a `YFile`, it would ensure that there is at least a new line character. But as it would be an opt-in, it's not too much opinionated.

Reply:
> Maybe this logic should live in jupyter_server_ydoc? We already [observe](https://github.com/jupyter-server/jupyter_server_ydoc/blob/d7e728b0c773977c6a307d7a8878b40c6be3a13d/jupyter_server_ydoc/ydoc.py#L171) document changes there, in order to auto-save after each change.

https://github.com/jupyter-server/jupyter_server_ydoc/pull/37#issuecomment-1266560042
I think we should keep it here in the `jupyter_ydoc`.

> If we keep it here in jupyter_ydoc, I think we should change the name: initialize() doesn't say much about what it's really doing. Maybe something like ensure_not_empty() would be better? For a YFile, it would ensure that there is at least a new line character. But as it would be an opt-in, it's not too much opinionated.

The idea of `initialize` is to be generic. When a document needs to do some specific operations after being initialized (like listening for changes in the cell list), the document knows that this method will be called after initializing the document.
I don't think it is a good idea to enforce a YFile containing at least a new line character.

Reply:
> The idea of `initialize` is to be generic.

But in this case it's not generic at all, it's adding a cell when there is no cell anymore.

> I don't think it is a good idea to enforce a YFile containing at least a new line character.

Why not? It's the equivalent of what you are doing with notebooks. And it's not enforced, as it would be an opt-in.

Reply:
> But in this case it's not generic at all, it's adding a cell when there is no cell anymore.

It is generic for the YBaseDoc. Then each document can overwrite `initialize` as needed.

> Why not? It's the equivalent of what you are doing with notebooks.

In the case of the notebook, we need to force a non-empty notebook due to a UX design. However, it is unnecessary for the YFile, so I would not do it because we are introducing unnecessary code. In addition, this will probably break the UI of the editor. The editor situates the cursor in the first line when the document is empty. What will happen if we don't have blank documents but documents with a line break? Will the editor always have two lines?

Reply:
To clarify, I don't think adding a new empty cell is a good idea either. However, as you can see in [this discussion](https://github.com/jupyterlab/jupyterlab/pull/13141) no one wants to break the actual UX and the best way of keeping it is by adding the cell here.

Reply:
> It is generic for the YBaseDoc. Then each document can overwrite `initialize` as needed.

Then why not overwriting `initialize` where you actually have that UX issue (in JupyterLab)? It seems to me that you are propagating a JupyterLab issue down to jupyter-ydoc, which is front-end agnostic.